src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs

```diff
@@ -170,14 +172,41 @@ public ConfidentialClientApplicationBuilder WithClientAssertion(string signedCli
         /// This is a delegate that computes a Base-64 encoded JWT for each authentication call.</param>
         /// <returns>The ConfidentialClientApplicationBuilder to chain more .With methods</returns>
         /// <remarks> Callers can use this mechanism to cache their assertions </remarks>
+        [Obsolete("Use WithClientAssertion(Func<CancellationToken, Task<string>> clientAssertionAsyncDelegate)")]
         public ConfidentialClientApplicationBuilder WithClientAssertion(Func<string> clientAssertionDelegate)
         {
             if (clientAssertionDelegate == null)
             {
                 throw new ArgumentNullException(nameof(clientAssertionDelegate));
             }
 
-            Config.SignedClientAssertionDelegate = clientAssertionDelegate;
+            Func<CancellationToken, Task<string>> clientAssertionAsyncDelegate = (token) =>
+            {                
+                return Task.FromResult(clientAssertionDelegate());
+            };
+
+
+            Config.SignedClientAssertionDelegate = clientAssertionAsyncDelegate;
+            Config.ConfidentialClientCredentialCount++;
+            return this;
+        }
+
+        /// <summary>
+        /// Configures an async delegate that creates a client assertion. See https://aka.ms/msal-net-client-assertion
```

Comment:
We'll have to add a paragraph in https://docs.microsoft.com/azure/active-directory/develop/msal-net-client-assertions

Reply:
PR added